### PR TITLE
merge flow-path into Flow

### PR DESCRIPTION
**Pull request by loganj**
- **Path merged into Flow**  
  Any object can be used as a key; a Path is a hierarchical key.
  Services can be configured for any key via a supplied factory.
  This makes Path essentially free to use.
- **All Context creation is internalized and managed by Flow.**  
  Setup/teardown is no longer the app author's concern.
- **Each key can now have an arbitrary state Bundle associated with it, in
  addition to ViewState.**
- **Keys are reference-counted and retained until all uses are gone.**  
  This enables apps to have multiple "live" keys, including Path keys
  with shared hierarchy prefixes. You can finally have a tree of scopes
  live in the app.
- **Key services are set up on demand, and torn down only when the
  traversal queue is completely flushed.**  
  This enables safe redirects without thrashing on scope setup/teardown.
- **Introduced KeyChanger interface and KeyDispatcher which uses it.**    
  KeyChanger is a simplified Dispatcher interface, intended for use by
  Dispatchers which only care about the keys at the top of the incoming
  and outgoing histories (most of them)

still TODO: samples, tests, docs

### CODE REVIEW

**Comment by loganj** [comment]
updated for copyright headers

**Comment by loganj** [comment]
updated to fix static imports of `checkNotNull`

File: flow/src/main/java/flow/DefaultKeyChanger.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+  @Override public void changeKey(@Nullable State outgoingState, State incomingState,
+      Flow.Direction direction, Context incomingStateContext,
+      Flow.TraversalCallback callback) {
+    if (textView == null) {
+      textView = new TextView(activity);
+      textView.setGravity(Gravity.CENTER);
+    }
+    textView.setText(incomingState.getKey().toString());
+    activity.setContentView(textView);
```
Comment (rjrjr) [inline flow/src/main/java/flow/DefaultKeyChanger.java]: Weird that you're calling setContentView over and over again.

File: flow/src/main/java/flow/Flow.java [hunk 2]
```diff
@@ -38,6 +40,19 @@ public static Flow get(Context context) {
...
+  public static <T> T getKey(Context context) {
+    return FlowContextWrapper.get(context).services.getKey();
+  }
+
+  public static <T> T getService(String serviceName, Context context) {
+    return FlowContextWrapper.get(context).services.getService(serviceName);
+  }
+
+
+  public static <T> T getKey(View view) {
```
Comment (rjrjr) [inline flow/src/main/java/flow/Flow.java]: Should group getKey() methods together

File: flow/src/main/java/flow/Flow.java [hunk 3]
```diff
@@ -81,11 +96,23 @@ public static boolean onNewIntent(Intent intent, Activity activity) {
...
       this.origin = from;
       this.destination = to;
       this.direction = direction;
+      this.keyManager = keyManager;
+    }
+
+    /**
+     * Creates a Context for the given key.
+     *
+     * Contexts can be created only for keys at the top of the origin and destination Histories.
```
Comment (rjrjr) [inline flow/src/main/java/flow/Flow.java]: Enforced?

File: flow/src/main/java/flow/Flow.java [hunk 4]
```diff
@@ -320,23 +350,31 @@ void enqueue(PendingTraversal pendingTraversal) {
...
+      // Is not set by noop and bootstrap transitions.
       if (nextHistory != null) {
+        tearDownKeys.add(history.top());
         history = nextHistory;
       }
       state = TraversalState.FINISHED;
       pendingTraversal = next;
-      if (dispatcher != null && pendingTraversal != null) {
+
+      if (pendingTraversal == null) {
```
Comment (rjrjr) [inline flow/src/main/java/flow/Flow.java]: :joy:

File: flow/src/main/java/flow/Flow.java (hunk 3 above)
Comment (loganj) [inline flow/src/main/java/flow/Flow.java]: enforced by findServices

File: flow/src/main/java/flow/DefaultKeyChanger.java (hunk 1 above)
Comment (loganj) [inline flow/src/main/java/flow/DefaultKeyChanger.java]: Yeah it is. Fixed.

File: flow/src/main/java/flow/Flow.java (hunk 2 above)
Comment (loganj) [inline flow/src/main/java/flow/Flow.java]: Done.

**Comment by loganj** [comment]
updated for feedback. @rjrjr PTAL.

File: flow/src/main/java/flow/InternalContextWrapper.java [hunk 5]
```diff
@@ -27,6 +53,11 @@ static Flow getFlow(Context context) {
         flow = InternalLifecycleIntegration.find(activity).flow;
       }
       return flow;
+    } else if (CONTEXT_MANAGER_SERVICE.equals(name)) {
+      if (keyManager == null) {
```
Comment (rjrjr) [inline flow/src/main/java/flow/InternalContextWrapper.java]: Why cache it?

Comment (loganj) [inline flow/src/main/java/flow/InternalContextWrapper.java]: Because I can, and it avoids future lookups through layers of ContextWrappers.

**Comment by rjrjr** [comment]
I don't have the big picture yet. Fine looking trees, unclear about the forest.

:shipit:

**Comment by loganj** [comment]
I'll definitely do a PR to review the forest when all the trees are more or less in place.

**Pull request merged**